### PR TITLE
refactor(ir): simplify expressions by not storing dtype and name

### PR DESCRIPTION
**Pull request by kszucs**
The point here is to not store any data in the expression classes but only the user facing API. This will enable us to keep a simpler operation hierarchy (without intermediare expressions) wrapped with a single top level user facing expression. 

A pretty small objective ended up as an invasive chage which highlights how coupled these module are. 

Though I managed to fix the majority of the issues, there are still multiple errors left.

### CODE REVIEW

File: ibis/backends/base/sql/registry/window.py [hunk 1]
```diff
@@ -95,7 +95,10 @@ def cumulative_to_window(translator, expr, window):
 
     klass = _cumulative_to_reduction[type(op)]
     new_op = klass(*op.args)
-    new_expr = expr._factory(new_op, name=expr._name)
+    new_expr = new_op.to_expr()
+
+    if (name := expr._safe_name) is not None:
```
Comment (kszucs) [inline ibis/backends/base/sql/registry/window.py:100]: Oooo, we can now use walrus?

Comment (cpcloud) [reply ibis/backends/base/sql/registry/window.py:100]: Yep, it was introduced in 3.8 and we're >= 3.8 :)

**Comment by cpcloud** [comment]
@kszucs Is this ready for review?

**Comment by kszucs** [comment]
@cpcloud should be ready for review

File: ibis/backends/base/sql/alchemy/registry.py [hunk 2]
```diff
@@ -332,7 +339,9 @@ def _cumulative_to_window(translator, expr, window):
 
     klass = _cumulative_to_reduction[type(op)]
     new_op = klass(*op.args)
-    new_expr = expr._factory(new_op, name=expr._name)
+    new_expr = type(expr)(new_op)
```
Comment (kszucs) [inline ibis/backends/base/sql/alchemy/registry.py:342]: Prefer `new_op.to_expr()`?

File: ibis/backends/base/sql/compiler/select_builder.py [hunk 3]
```diff
@@ -555,10 +559,12 @@ def _visit_filter(self, expr):
...
             right = self._visit_filter(op.right)
             unchanged = left is op.left and right is op.right
-            if not unchanged:
-                return expr._factory(type(op)(left, right))
-            else:
+            if unchanged:
                 return expr
+            else:
+                new_op = type(op)(left, right)
+                new_expr = type(expr)(new_op)
```
Comment (kszucs) [inline ibis/backends/base/sql/compiler/select_builder.py:566]: Try `new_op.to_expr()`

File: ibis/backends/base/sql/compiler/select_builder.py [hunk 4]
```diff
@@ -479,6 +479,7 @@ def _visit_select_expr(self, expr):
                     new_args.append(arg)
 
             if not unchanged:
+                # TODO(kszucs): remove it
                 return expr._factory(type(op)(*new_args))
```
Comment (kszucs) [inline ibis/backends/base/sql/compiler/select_builder.py:483]: Remove `Expr._factory`

File: ibis/backends/base/sql/compiler/translator.py [hunk 5]
```diff
@@ -287,6 +288,7 @@ def decorator(f):
 rewrites = ExprTranslator.rewrites
 
 
+# TODO(kszucs): is this used?
```
Comment (kszucs) [inline ibis/backends/base/sql/compiler/translator.py:291]: Remove TODO note

File: ibis/backends/clickhouse/tests/test_functions.py [hunk 6]
```diff
@@ -328,13 +328,14 @@ def test_parse_url_query_parameter(con, translate):
...
         (L('foobar').find('bar'), 3),
         (L('foobar').find('baz'), -1),
-        (L('foobar').like('%bar'), True),
-        (L('foobar').like('foo%'), True),
-        (L('foobar').like('%baz%'), False),
-        (L('foobar').like(['%bar']), True),
-        (L('foobar').like(['foo%']), True),
-        (L('foobar').like(['%baz%']), False),
-        (L('foobar').like(['%bar', 'foo%']), True),
+        # FIXME(kszucs)
```
Comment (kszucs) [inline ibis/backends/clickhouse/tests/test_functions.py:331]: Uncomment these cases

File: ibis/expr/datatypes/core.py [hunk 7]
```diff
@@ -131,6 +131,7 @@ def equals(self, other):
             )
         return super().__cached_equals__(other)
 
+    # TODO(kszucs): remove it
     def _factory(self, nullable: bool = True) -> DataType:
```
Comment (kszucs) [inline ibis/expr/datatypes/core.py:135]: Try to remove `DataType._factory`

File: ibis/expr/format.py [hunk 8]
```diff
@@ -559,6 +559,13 @@ def _fmt_value_sort_key(op: ops.SortKey, *, aliases: Aliases) -> str:
     return f"{sort_direction}|{expr}"
 
 
+@fmt_value.register
+def _fmt_topk(op: ops.TopK, *, aliases: Aliases) -> str:
+    arg = fmt_value(op.arg, aliases=aliases)
+    by = fmt_value(op.by, aliases=aliases)
+    return f"TopK({arg}, {op.k}, {by})"
```
Comment (kszucs) [inline ibis/expr/format.py:566]: @cpcloud just quickly added an implementation since I changed the base class of TopK from ValueOp to Node.

File: ibis/expr/operations/analytic.py [hunk 9]
```diff
@@ -7,18 +7,16 @@
...
 
-@public
-class AnalyticOp(ValueOp):
-    pass
-
-
 @public
 class WindowOp(ValueOp):
+    # TODO(kszucs): remove this property
+    display_argnames = False
```
Comment (kszucs) [inline ibis/expr/operations/analytic.py:13]: @cpcloud is `display_argnames` is still used in the new repr?

File: ibis/expr/operations/analytic.py [hunk 10]
```diff
@@ -219,34 +243,28 @@ class Any(ValueOp):
...
 
+    # TODO(kszucs): we should probably remove the negate methods
     def negate(self):
         return NotAny(self.arg)
 
 
 @public
 class NotAny(Any):
+
+    # TODO(kszucs): we should probably remove the negate methods
```
Comment (kszucs) [inline ibis/expr/operations/analytic.py:268]: Create a follow-up for this.

File: ibis/expr/operations/arrays.py [hunk 11]
```diff
@@ -18,40 +20,50 @@ def __init__(self, cols):
...
 
 
 @public
 class ArrayLength(UnaryOp):
     arg = rlz.array
-    output_type = rlz.shape_like('arg', dt.int64)
+
+    # TODO(kszucs): output shape should be defined by UnaryUp already
+    output_dtype = dt.int64
+    output_shape = rlz.shape_like("arg")
```
Comment (kszucs) [inline ibis/expr/operations/arrays.py:35]: Remove `output_shape` definition here.

File: ibis/expr/operations/core.py [hunk 12]
```diff
@@ -114,14 +117,80 @@ def flat_args(self):
 
 @public
 class ValueOp(Node):
+    # columnar if any of the arguments is column-like
+    output_shape = rlz.shape_like("args")
+
+    # TODO(kszucs): we may want to enable this by default instead of
```
Comment (kszucs) [inline ibis/expr/operations/core.py:123]: Since operations are now immutable we can materialise `output_shape` and `output_dtype` on construction. Should experiment with this.

Eventually we should materialise `output_type` for all `Node` objects on construction.

File: ibis/expr/operations/core.py [hunk 13]
```diff
@@ -114,14 +117,80 @@ def flat_args(self):
...
+    # def resolve_name(self):
+    #     raise ExpressionError(f'Expression is not named: {type(self)}')
+
+    @property
+    @abstractmethod
+    def output_dtype(self):
+        ...
+
+    @property
+    def output_type(self):
```
Comment (kszucs) [inline ibis/expr/operations/core.py:140]: Add comments explaining that we defer the type selection to the `output_dtype` and `output_shape` properties.

File: ibis/expr/operations/generic.py [hunk 14]
```diff
@@ -107,30 +108,43 @@ class TableArrayView(ValueOp):
...
+    output_dtype = property(attrgetter("to"))
+
     # see #396 for the issue preventing an implementation of resolve_name
 
-    def output_type(self):
-        return rlz.shape_like(self.arg, dtype=self.to)
+
+# TODO(kszucs): find a better name
+# @public
+# class ReinterpretCast(ValueOp):
```
Comment (kszucs) [inline ibis/expr/operations/generic.py:137]: Remove this.

File: ibis/expr/operations/generic.py [hunk 15]
```diff
@@ -156,25 +170,28 @@ class NotNull(UnaryOp):
...
     case().when(expr.notnull(), expr)
           .else_(null_substitute_expr)
     """
 
     arg = rlz.any
     ifnull_expr = rlz.any
-    output_type = rlz.shape_like('args')
+
+    # we may not need this if we use this rule on valueop by default
+    output_dtype = rlz.dtype_like("args")
```
Comment (kszucs) [inline ibis/expr/operations/generic.py:194]: Remove this since ValueOp's default `output_dtype` is identical with this property.

File: ibis/expr/operations/generic.py [hunk 16]
```diff
@@ -183,7 +200,9 @@ class NullIf(ValueOp):
 
     arg = rlz.any
     null_if_expr = rlz.any
-    output_type = rlz.shape_like('args')
+    output_dtype = rlz.dtype_like(
```
Comment (kszucs) [inline ibis/expr/operations/generic.py:203]: Remove this.

File: ibis/backends/base/sql/compiler/translator.py [hunk 17]
```diff
@@ -200,6 +201,27 @@ def __init__(self, expr, context, named=False, permit_subquery=False):
...
+    def _needs_name(self, expr):
+        if not self.named:
+            return False
+
+        op = expr.op()
+        if isinstance(op, ops.TableColumn):
+            # This column has been given an explicitly different name
+            if expr.get_name() != op.name:
+                return True
+            return False
```
Comment (cpcloud) [inline ibis/backends/base/sql/compiler/translator.py:213]:
```suggestion
            return expr.get_name() != op.name
```

File: ibis/backends/base/sql/compiler/translator.py [hunk 18]
```diff
@@ -200,6 +201,27 @@ def __init__(self, expr, context, named=False, permit_subquery=False):
...
+
+        if expr.get_name() is unnamed:
+            return False
+
+        return True
```
Comment (cpcloud) [inline ibis/backends/base/sql/compiler/translator.py:218]:
```suggestion
        return expr.get_name() is not unnamed
```

File: ibis/backends/pandas/execution/selection.py [hunk 19]
```diff
@@ -97,7 +97,10 @@ def compute_projection_column_expr(
     timecontext: Optional[TimeContext],
     **kwargs,
 ):
-    result_name = getattr(expr, '_name', None)
+    if expr.has_name():
+        result_name = expr.get_name()
+    else:
+        result_name = None
```
Comment (cpcloud) [inline ibis/backends/pandas/execution/selection.py:103]:
```suggestion
    result_name = expr._safe_name
```

File: ibis/backends/tests/test_timecontext.py [hunk 20]
```diff
@@ -61,7 +61,7 @@ def ctx_col():
             ibis.trailing_window(ibis.interval(days=3), order_by=ORDER_BY_COL),
             id="order_by",
         ),
-        param(
+        pytest.param(
```
Comment (cpcloud) [inline ibis/backends/tests/test_timecontext.py:64]:
```suggestion
        param(
```

File: ibis/expr/operations/core.py [hunk 21]
```diff
@@ -107,14 +110,44 @@ def flat_args(self):
...
         return distinct_roots(*self.exprs)
 
-    def resolve_name(self):
-        raise ExpressionError(f'Expression is not named: {type(self)}')
+    @property
+    @abstractmethod
+    def output_dtype(self):
+        ...
+
+    # TODO(kszucs): use immutable_property instead
```
Comment (cpcloud) [inline ibis/expr/operations/core.py:129]: Should this be addressed?

File: ibis/expr/operations/generic.py (hunk 15 above)
Comment (cpcloud) [reply ibis/expr/operations/generic.py:184]: Bump

File: ibis/expr/operations/generic.py (hunk 16 above)
Comment (cpcloud) [reply ibis/expr/operations/generic.py:193]: Bump

File: ibis/expr/operations/relations.py [hunk 22]
```diff
@@ -71,6 +71,13 @@ class UnboundTable(PhysicalTable):
     schema = rlz.instance_of(sch.Schema)
     name = rlz.optional(rlz.instance_of(str), default=genname)
 
+    # TODO(kszucs): something like a HasName trait could be useful here
```
Comment (cpcloud) [inline ibis/expr/operations/relations.py:74]: Isn't `HasName` implemented elsewhere in this PR?

File: ibis/expr/types/core.py [hunk 23]
```diff
@@ -24,6 +24,10 @@
     from .generic import ValueExpr
 
 
+# TODO(kszucs): value exprs should expose a dtype() function
+# TODO(kszucs): table exprs should expose a schema() function
```
Comment (cpcloud) [inline ibis/expr/types/core.py:28]: `TableExpr` already exposes a `schema` method, and `ValueExpr` has the `type()` method.

File: ibis/expr/types/relations.py [hunk 24]
```diff
@@ -74,14 +75,7 @@ def f(
...
 @public
-class TableExpr(Expr):
-    @property
-    def _factory(self):
-        def factory(arg):
-            return TableExpr(arg)
-
-        return factory
-
+class TableExpr(Expr, HasName):
```
Comment (cpcloud) [inline ibis/expr/types/relations.py:78]: What does `TableExpr.name` compile to? Something like `(SELECT * FROM t) name`?

File: ibis/tests/expr/test_format.py [hunk 25]
```diff
@@ -199,14 +198,6 @@ def test_same_column_multiple_aliases():
     assert "fakealias2: r0.col" in result
 
 
-def test_scalar_parameter_repr():
```
Comment (cpcloud) [inline ibis/tests/expr/test_format.py:202]: Why did you delete this test?

File: ibis/tests/expr/test_temporal.py [hunk 26]
```diff
@@ -78,7 +78,7 @@ def test_multiply(expr):
     'expr',
     [
         api.interval(days=1) + api.interval(days=1),
-        api.interval(days=2) + api.interval(hours=4),
+        # api.interval(days=2) + api.interval(hours=4),
```
Comment (cpcloud) [inline ibis/tests/expr/test_temporal.py:81]: Should these tests be reenabled?

File: ibis/tests/expr/test_temporal.py [hunk 27]
```diff
@@ -90,7 +90,7 @@ def test_add(expr):
     'expr',
     [
         api.interval(days=3) - api.interval(days=1),
-        api.interval(days=2) - api.interval(hours=4),
+        # api.interval(days=2) - api.interval(hours=4),
```
Comment (cpcloud) [inline ibis/tests/expr/test_temporal.py:93]: Reenable?

File: ibis/tests/expr/test_value_exprs.py [hunk 28]
```diff
@@ -515,12 +517,12 @@ def test_number_to_string_scalar():
...
-    expr = table.f.cast('string')
-    assert expr.get_name() == 'cast(f, string)'
+# def test_casted_exprs_are_named(table):
+#     expr = table.f.cast('string')
+#     assert expr.get_name() == 'cast(f, string)'
 
-    # it works! per GH #396
-    expr.value_counts()
+#     # it works! per GH #396
+#     expr.value_counts()
```
Comment (cpcloud) [inline ibis/tests/expr/test_value_exprs.py:525]: Should this test be uncommented?

**Review by cpcloud — CHANGES_REQUESTED** [review]

File: ibis/tests/expr/test_format.py (hunk 25 above)
Comment (kszucs) [reply ibis/tests/expr/test_format.py:202]: Restored.

File: ibis/tests/expr/test_temporal.py (hunk 26 above)
Comment (kszucs) [reply ibis/tests/expr/test_temporal.py:81]: Reenabled.

File: ibis/tests/expr/test_temporal.py (hunk 27 above)
Comment (kszucs) [reply ibis/tests/expr/test_temporal.py:93]: Reenabled.

File: ibis/tests/expr/test_value_exprs.py (hunk 28 above)
Comment (kszucs) [reply ibis/tests/expr/test_value_exprs.py:525]: Reenabled.

File: ibis/expr/types/core.py (hunk 23 above)
Comment (kszucs) [reply ibis/expr/types/core.py:28]: Removed comments.

File: ibis/expr/operations/core.py (hunk 13 above)
Comment (cpcloud) [reply ibis/expr/operations/core.py:131]: Bump

File: ibis/expr/operations/core.py (hunk 21 above)
Comment (cpcloud) [reply ibis/expr/operations/core.py:129]: Bump

File: ibis/expr/types/relations.py (hunk 24 above)
Comment (cpcloud) [reply ibis/expr/types/relations.py:78]: Bump

File: ibis/expr/operations/analytic.py (hunk 10 above)
Comment (kszucs) [reply ibis/expr/operations/analytic.py:265]: Removed the comment.

File: ibis/expr/operations/core.py (hunk 12 above)
Comment (kszucs) [reply ibis/expr/operations/core.py]: Updated.

File: ibis/expr/operations/core.py (hunk 13 above)
Comment (kszucs) [reply ibis/expr/operations/core.py:142]: Added, but not extensively.

File: ibis/expr/operations/generic.py (hunk 15 above)
Comment (kszucs) [reply ibis/expr/operations/generic.py:184]: Updated.

File: ibis/expr/operations/generic.py (hunk 16 above)
Comment (kszucs) [reply ibis/expr/operations/generic.py]: Updated.

File: ibis/expr/operations/core.py (hunk 21 above)
Comment (kszucs) [reply ibis/expr/operations/core.py]: Added.

File: ibis/expr/operations/relations.py (hunk 22 above)
Comment (kszucs) [reply ibis/expr/operations/relations.py:74]: Removed the comment. Going to take a look later.

File: ibis/expr/types/relations.py (hunk 24 above)
Comment (kszucs) [reply ibis/expr/types/relations.py:78]: What do you mean under compile to? `TableExpr.name` is a method which instantiates the `ops.Alias()` objects.

**Comment by kszucs** [comment]
This should be ready now. @cpcloud could you please take a look?

File: ibis/expr/types/relations.py (hunk 24 above)
Comment (cpcloud) [reply ibis/expr/types/relations.py:78]: Right, this PR introduces a `name` on `TableExpr`s that doesn't currently exist. What code does the new `ops.Alias` translate to?

**Comment by cpcloud** [comment]
@kszucs This looks good to me. Can you squash and make sure to add at least one commit with a nice `BREAKING CHANGE` section indicating what users may need to do to update their code to work with 3.0.0?

File: ibis/backends/base/sql/compiler/translator.py [hunk 29]
```diff
@@ -345,7 +345,11 @@ def _category_label(expr):
     if op.nulls is not None:
         stmt = stmt.else_(op.nulls)
 
-    return stmt.end().name(expr._name)
+    result = stmt.end()
+    if expr.has_name():
+        result = result.name(expr.get_name())
```
Comment (cpcloud) [inline ibis/backends/base/sql/compiler/translator.py:350]: I guess we don't have a test that hits this branch. Probably fine to get rid of the conditional.

File: ibis/backends/base/sql/compiler/translator.py [hunk 30]
```diff
@@ -331,7 +327,11 @@ def _bucket(expr):
         stmt = stmt.when(cmp(op.buckets[-1], op.arg), bucket_id)
         bucket_id += 1
 
-    return stmt.end().name(expr._name)
+    result = stmt.end()
+    if expr.has_name():
+        result = result.name(expr.get_name())
```
Comment (cpcloud) [inline ibis/backends/base/sql/compiler/translator.py:332]: Looks like another place where we have new code that isn't tested.

File: ibis/backends/base/sql/compiler/select_builder.py [hunk 31]
```diff
@@ -500,8 +501,11 @@ def _visit_select_Histogram(self, expr):
             binwidth = op.binwidth
             base = op.base
 
-        bucket = (op.arg - base) / binwidth
-        return bucket.floor().name(expr._name)
+        bucket = ((op.arg - base) / binwidth).floor()
+        if expr.has_name():
+            bucket = bucket.name(expr.get_name())
```
Comment (cpcloud) [inline ibis/backends/base/sql/compiler/select_builder.py:506]: It looks like there are a bunch of these `has_name` checks that aren't tested.

File: ibis/backends/base/sql/alchemy/translator.py [hunk 32]
```diff
@@ -38,7 +38,8 @@ class AlchemyExprTranslator(ExprTranslator):
     def name(self, translated, name, force=True):
         if hasattr(translated, 'label'):
             return translated.label(name)
-        return translated
+        else:
+            return translated
```
Comment (cpcloud) [inline ibis/backends/base/sql/alchemy/translator.py:42]: No real reason to change this line. It's not covered and it's yet another file to review.

File: ibis/backends/base/sql/alchemy/registry.py [hunk 33]
```diff
@@ -327,7 +334,9 @@ def _cumulative_to_window(translator, expr, window):
 
     klass = _cumulative_to_reduction[type(op)]
     new_op = klass(*op.args)
-    new_expr = expr._factory(new_op, name=expr._name)
+    new_expr = new_op.to_expr()
+    if expr.has_name():
+        new_expr = new_expr.name(expr.get_name())
```
Comment (cpcloud) [inline ibis/backends/base/sql/alchemy/registry.py:339]: Another case of `has_name` that does have a test.

File: ibis/expr/analysis.py [hunk 34]
```diff
@@ -145,10 +145,9 @@ def _visit(self, expr):
             subbed_args.append(subbed_arg)
 
         subbed_node = type(node)(*subbed_args)
-        if isinstance(expr, ir.ValueExpr):
-            result = expr._factory(subbed_node, name=expr._name)
-        else:
-            result = expr._factory(subbed_node)
+        result = type(expr)(subbed_node)
+        if isinstance(expr, ir.ValueExpr) and expr.has_name():
+            result = result.name(name=expr.get_name())
```
Comment (cpcloud) [inline ibis/expr/analysis.py:150]: More has_names without tests :)

**Review by cpcloud — APPROVED** [review]
No big deal, but looks like there are a few places where we're checking `has_name()` that don't have tests. We can address in a follow-up.

File: ibis/backends/base/sql/alchemy/translator.py (hunk 32 above)
Comment (kszucs) [reply ibis/backends/base/sql/alchemy/translator.py]: Removed that branch.

File: ibis/backends/base/sql/alchemy/registry.py (hunk 33 above)
Comment (kszucs) [reply ibis/backends/base/sql/alchemy/registry.py:339]: All 4 `ibis/backends/tests/test_window.py::test_grouped_bounded_expanding_window` tests fail if I remove the `has_name()` check.

File: ibis/backends/base/sql/compiler/select_builder.py (hunk 31 above)
Comment (kszucs) [reply ibis/backends/base/sql/compiler/select_builder.py:506]: Breaks tests if I remove the `has_name()` check.

File: ibis/backends/base/sql/compiler/translator.py (hunk 30 above)
Comment (kszucs) [reply ibis/backends/base/sql/compiler/translator.py:332]: Breaks tests if I remove the has_name() check.

File: ibis/backends/base/sql/compiler/translator.py (hunk 29 above)
Comment (kszucs) [reply ibis/backends/base/sql/compiler/translator.py:350]: Breaks tests if I remove the has_name() check.

File: ibis/expr/analysis.py (hunk 34 above)
Comment (kszucs) [reply ibis/expr/analysis.py:150]: Removed the `has_name()` check.

File: ibis/backends/base/sql/compiler/translator.py (hunk 29 above)
Comment (kszucs) [reply ibis/backends/base/sql/compiler/translator.py:350]: Does the postgres build report to codecov?

**Comment by cpcloud** [comment]
I think there might be a bit of a race condition with code cov uploading, I'll investigate this week.

**Comment by kszucs** [comment]
Squashed and listed the breaking changes. Apparently the commit message got too verbose, can we relax that condition?

File: ibis/expr/types/relations.py (hunk 24 above)
Comment (kszucs) [reply ibis/expr/types/relations.py]: Now I managed to understand :)

Indeed name() hasn't been exposed on TableExpr before, so I removed the HasName trait and defined the has_name(), get_name() methods on Expr base class and added name() method to ValueExpr.

**Comment by cpcloud** [comment]
> Squashed and listed the breaking changes. Apparently the commit message got too verbose, can we relax that condition?

In theory wrapping the lines should be enough.

**Comment by kszucs** [comment]
> > Squashed and listed the breaking changes. Apparently the commit message got too verbose, can we relax that condition?
> 
> In theory wrapping the lines should be enough.

Right, didn't read the error message carefully.

**Comment by kszucs** [comment]
@cpcloud all builds are green now
